crates/re_space_view_spatial/src/scene/parts/meshes.rs

```diff
@@ -0,0 +1,119 @@
+use re_components::{ColorRGBA, Component, InstanceKey, Mesh3D};
+use re_data_store::EntityPath;
+use re_query::{EntityView, QueryError};
+use re_renderer::renderer::MeshInstance;
+use re_viewer_context::{
+    ArchetypeDefinition, DefaultColor, ScenePart, SceneQuery, SpaceViewHighlights, ViewerContext,
+};
+
+use crate::instance_hash_conversions::picking_layer_id_from_instance_path_hash;
+use crate::mesh_cache::MeshCache;
+use crate::scene::{
+    contexts::{SpatialSceneContext, SpatialSceneEntityContext},
+    parts::entity_iterator::process_entity_views,
+};
+use crate::SpatialSpaceViewClass;
+
+use super::{instance_path_hash_for_picking, SpatialScenePartData, SpatialSpaceViewState};
+
+#[derive(Default)]
+pub struct MeshPart(SpatialScenePartData);
+
+impl MeshPart {
+    fn process_entity_view(
+        &mut self,
+        ctx: &mut ViewerContext<'_>,
+        instances: &mut Vec<MeshInstance>,
+        ent_view: &EntityView<Mesh3D>,
+        ent_path: &EntityPath,
+        ent_context: &SpatialSceneEntityContext<'_>,
+    ) -> Result<(), QueryError> {
+        let _default_color = DefaultColor::EntityPath(ent_path);
+
+        let visitor = |instance_key: InstanceKey,
+                       mesh: re_components::Mesh3D,
+                       _color: Option<ColorRGBA>| {
+            let picking_instance_hash = instance_path_hash_for_picking(
+                ent_path,
+                instance_key,
+                ent_view.num_instances(),
+                ent_context.highlight.any_selection_highlight,
+            );
+
+            let outline_mask_ids = ent_context.highlight.index_outline_mask(instance_key);
+
+            if let Some(mesh) =
+                ctx.cache
+                    .entry::<MeshCache>()
+                    .entry(&ent_path.to_string(), &mesh, ctx.render_ctx)
+            {
+                instances.extend(mesh.mesh_instances.iter().map(move |mesh_instance| {
+                    MeshInstance {
+                        gpu_mesh: mesh_instance.gpu_mesh.clone(),
+                        world_from_mesh: ent_context.world_from_obj * mesh_instance.world_from_mesh,
+                        outline_mask_ids,
+                        picking_layer_id: picking_layer_id_from_instance_path_hash(
+                            picking_instance_hash,
+                        ),
+                        ..Default::default()
+                    }
+                }));
+
+                self.0
+                    .extend_bounding_box(*mesh.bbox(), ent_context.world_from_obj);
+            };
+        };
+
+        ent_view.visit2(visitor)?;
+
+        Ok(())
+    }
+}
+
+impl ScenePart<SpatialSpaceViewClass> for MeshPart {
+    fn archetype(&self) -> ArchetypeDefinition {
+        vec1::vec1![Mesh3D::name(), InstanceKey::name(), ColorRGBA::name()]
```

Comment:
Adding `InstanceKey` for all archetypes seems a bit weird/error prone. What does it mean for it to be missing?

Reply:
Agreed, this came up in the past as well. InstanceKey is oddly special and needed for anything that isn't a mono query. Don' think that's actionable in this PR. It's more a question for @teh-cmc 's archetype definitions and the facade work.